hrm-domain/lambdas/custom-integrations/uscr-beacon-poller/src/apiChunkReader.ts

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright (C) 2021-2023 Technology Matters
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see https://www.gnu.org/licenses/.
+ */
+
+import { getSsmParameter, putSsmParameter } from '@tech-matters/ssm-cache';
+import { isErr } from '@tech-matters/types';
+import { ItemProcessor } from './types';
+
+type ChunkReaderConfig<TItem> = {
+  url: URL;
+  headers: Record<string, string>;
+  lastUpdateSeenSsmKey: string;
+  itemProcessor: ItemProcessor<TItem>;
+  maxItemsInChunk: number;
+  maxChunksToRead: number;
+  itemTypeName?: string; // Just for logging
+};
+
+const processChunk = async <TItem>(
+  items: TItem[],
+  lastSeen: string,
+  itemProcessor: ItemProcessor<TItem>,
+  itemTypeName: string = 'item',
+): Promise<string> => {
+  let updatedLastSeen = lastSeen;
+  for (const item of items) {
+    console.debug(`Start processing ${itemTypeName}:`, lastSeen);
+    const processorResult = await itemProcessor(item);
+    if (isErr(processorResult)) {
+      console[processorResult.error.level](
+        processorResult.message,
+        processorResult.error,
+      );
+      updatedLastSeen = processorResult.error.lastUpdated ?? updatedLastSeen;
+    } else {
+      updatedLastSeen = processorResult.unwrap();
+    }
+  }
+  return updatedLastSeen;
+};
+
+export const readApiInChunks = async <TItem>({
+  url,
+  headers,
+  lastUpdateSeenSsmKey,
+  maxItemsInChunk,
+  maxChunksToRead,
+  itemProcessor,
+  itemTypeName = 'item',
+}: ChunkReaderConfig<TItem>) => {
+  let lastUpdateSeen = await getSsmParameter(lastUpdateSeenSsmKey);
+  let processedAllItems = false;
+  for (let i = 0; i < maxChunksToRead; i++) {
+    console.info(`Last ${itemTypeName} update before:`, lastUpdateSeen);
+    // Query Beacon API
+    url.searchParams.set('updatedAfter', lastUpdateSeen);
+    url.searchParams.set('max', maxItemsInChunk.toString());
+    console.info('Querying:', url);
+    const response = await fetch(url, { headers });
+    console.debug(`Beacon ${itemTypeName} API responded with status:`, response.status);
+    if (response.ok) {
```

Comment:
Do we want any more specific error handling if the response isn't ok, rather than the warning on L111 that we didn't get everything?

It also seems like if there's any kind of error, this is just going to keep retrying `maxChunksToRead` times, which may not be the right thing. If this is a necessary expedient for now, fine.

Reply:
I was going to add some logs tailored to the alarms and dashboards we are going to set up in my current ticker CHI-3241

I don't know what else we want to do for the retries, given we are the only client and exponential backoff doesn't seem warranted. In it's current state it will just keep retrying up to the maximum amount we had specified for calling the API, which seems reasonable to me for a starter